lib/scala/kg/src/main/scala/io/github/tetherlessworld/mcsapps/lib/kg/stores/postgres/PostgresKgCommandStore.scala

```diff
@@ -107,6 +110,159 @@ class PostgresKgCommandStore @Inject()(configProvider: PostgresStoreConfigProvid
     override final def putSources(sources: Iterator[KgSource]): Unit =
       runSyncTransaction(DBIO.sequence(batchedSourceInserts(sources)))
 
+    private def writeNodePageRank(convergenceThreshold: Double = 0.0000001, dampingFactor: Double = 0.85, maxIterations: Int = 20): Unit = {
+      val numNodesAction = nodes.size.result
+//      Initialize node page ranks to 1/n where n is the number of nodes
+      val initializeNodePageRanksAction = numNodesAction.flatMap {
+        case (numNodes) => nodes.map(_.pageRank).update(Some((1.0 / numNodes).toFloat))
+      }
+//      Initialize node in/out degrees and page ranks
+      val numNodes = runSyncTransaction(DBIO.sequence(List(initializeNodePageRanksAction, writeNodeDegreesAction)))(0)
+
+      if (numNodes == 0) {
+        return
+      }
+
+      for (_ <- 1 to maxIterations) {
+//        intellij shows writeNodePageRankAction returning a List[Double] but for some
+//         reason sbt compiles it as returning List[Any] so forced to convert to Double
+//        Update node page ranks and retrieve page rank delta value
+        val delta = runSyncTransaction(writeNodePageRankAction(dampingFactor.toFloat))(1).asInstanceOf[Number].doubleValue
+
+//        If page rank has converged, exit
+        if (delta < convergenceThreshold) {
+          return
+        }
+      }
+    }
+
+    private def writeNodeLabelPageRank(convergenceThreshold: Double = 0.0000001, dampingFactor: Double = 0.85, maxIterations: Int = 20): Unit = {
+      val numNodeLabelsAction = nodeLabels.size.result
+      val initializeNodeLabelPageRanksAction = numNodeLabelsAction.flatMap {
+        case (numNodeLabels) => nodeLabels.map(_.pageRank).update(Some((1.0 / numNodeLabels).toFloat))
+      }
+      val numNodes = runSyncTransaction(DBIO.sequence(List(initializeNodeLabelPageRanksAction, writeNodeLabelDegreesAction)))(0)
+
+      if (numNodes == 0) {
+        return
+      }
+
+      for (_ <- 1 to maxIterations) {
+        val delta = runSyncTransaction(writeNodeLabelPageRankAction(dampingFactor.toFloat))(1).asInstanceOf[Number].doubleValue
+
+        if (delta < convergenceThreshold) {
+          return
+        }
+      }
+    }
+
+    private def writeNodePageRankAction(dampingFactor: Float) = {
+      val temporaryTableName = "temp_node_page_rank"
+      DBIO.sequence(List(
+        // Calculates new page rank for each node and saves in a temporary table
+        sqlu"""
+            SELECT n.id as node_id, ${dampingFactor} * sum(n_neighbor.page_rank / n.out_degree) + (1 - ${dampingFactor}) as page_rank
+            INTO #$temporaryTableName
+            FROM node n
+            JOIN edge e
+            ON e.object_node_id = n.id
+            JOIN node n_neighbor
+            ON n_neighbor.id = e.subject_node_id
+            GROUP BY n.id
+          """,
+        // Check if page rank has converged by calculating the difference
+        //  between the new page ranks and the current page ranks
+        sql"""
```

Comment:
I don't know Slick well enough to know.. if this is the only result, does it return it from `DBIO.sequence`? You should declare the method return type on `writeNodePageRankAction`, in any case.

Reply:
So this part is a little ugly, `DBIO.sequence` returns a `List[<Result type]`. The `sqlu""` statements' return type is `Int` while the `sql""` can be cast to a specific type, in this case `Double`. So in this case, the return type will end up being `List[Int, Double, Int, Int]`. I've declared the return type as `List[Any]`.